lesson34/src/js/article.js

```diff
@@ -0,0 +1,159 @@
+  const favoriteArticlesDataOfLoginUser = favoriteArticlesData[loginUserToken] ?? [];
+  const registeredFavoriteArticleData = [
+    ...favoriteArticlesDataOfLoginUser,
+    {
+      [currentArticleData.id]: {
+        "title": currentArticleData.title,
+        "date": currentArticleData.date,
+        "thumbnail": currentArticleData.thumbnail
+      }
+    }
+  ]
+  const margeData = { ...favoriteArticlesData, [loginUserToken]: registeredFavoriteArticleData };
+  localStorage.setItem("favoriteArticles", JSON.stringify(margeData));
+}
+
+const renderArticle = (data) => {
+  const articleFragment = document.createDocumentFragment();
+  const article = createElementWithClassName("article", "article");
+  const articleTitle = createElementWithClassName("h1", "article__title");
+  const articleContent = createElementWithClassName("div", "article__content");
+  articleTitle.textContent = data.title;
+  articleContent.textContent = data.content;
+  articleFragment.appendChild(articleTitle).after(createArticleMeta(data));
+  data.thumbnail && articleFragment.appendChild(createArticleThumbnail(data));
+  articleFragment.appendChild(articleContent);
+  articleWrapper.appendChild(article).appendChild(articleFragment);
+}
+
+const addEventListenerForFavoriteButton = () => {
+  const favoriteButton = document.getElementById("js-favorite-button");
+  favoriteButton.addEventListener("click", (event) => {
+    event.preventDefault();
+    registerFavoriteButton();
+    setFavoriteArticleDataInLocalStorage();
+  });
+}
+
+const renderFavoriteButton = () => {
+  const favoriteButton = createElementWithClassName("button", "article__favorite-icon");
+  const img = document.createElement("img");
+  img.src = "./img/star-icon.svg";
+  favoriteButton.dataset.favoriteAction = "unregistered";
+  favoriteButton.id = "js-favorite-button";
+  articleWrapper.appendChild(favoriteButton).appendChild(img);
+}
+
+const createArticleMeta = ({ date }) => {
+  const articleMetaWrapper = createElementWithClassName("div", "article__meta");
+  const articleDate = createElementWithClassName("p", "article__date");
+  articleDate.textContent = date;
+  articleMetaWrapper.appendChild(articleDate);
+  return articleMetaWrapper;
+}
+
+const createArticleThumbnail = ({ thumbnail }) => {
+  const articleThumbnail = createElementWithClassName("p", "article__thumbnail");
+  const img = document.createElement("img");
+  img.src = thumbnail;
+  articleThumbnail.append(img);
+  return articleThumbnail;
+}
+
+const findArticleData = (data) => {
+  const urlParams = new URLSearchParams(window.location.search);
+  let result;
+  data.forEach(({ article }) => {
+    article.forEach((data) => data.id === urlParams.get("id") && (result = data));
+  });
+  return result;
+}
+
+const isRegisteredFavoriteArticle = () => {
+  const loginUserToken = localStorage.getItem("token");
+  const userFavoriteArticlesData = JSON.parse(localStorage.getItem("favoriteArticles"))?.[loginUserToken];
```

Comment:
This is [Optional chaining ](https://developer.mozilla.org/ja/docs/Web/JavaScript/Reference/Operators/Optional_chaining).
Accessing a non-existent key usually results in an error, but by using Optional chaining, "undefined" is returned instead of an error.
In this case, since the access destination is not guaranteed, a crash is avoided by using Optional Chaining.

Reply:
I learned a lot from all the syntax I've never seen before. Thank you very much. Sae-san, who is mastering this, is amazing!!👏